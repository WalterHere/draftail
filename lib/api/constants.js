import { Map } from 'immutable';
import { DefaultDraftBlockRenderMap } from 'draft-js';

// Originally from https://github.com/draft-js-utils/draft-js-utils/blob/master/src/Constants.js.
export const BLOCK_TYPE = {
    // This is used to represent a normal text block (paragraph).
    UNSTYLED: 'unstyled',
    HEADER_ONE: 'header-one',
    HEADER_TWO: 'header-two',
    HEADER_THREE: 'header-three',
    HEADER_FOUR: 'header-four',
    HEADER_FIVE: 'header-five',
    HEADER_SIX: 'header-six',
    UNORDERED_LIST_ITEM: 'unordered-list-item',
    ORDERED_LIST_ITEM: 'ordered-list-item',
    BLOCKQUOTE: 'blockquote',
    PULLQUOTE: 'pullquote',
    CODE: 'code-block',
    ATOMIC: 'atomic',
    // HR block.
    BREAK: 'atomic:break',
};

// Originally from https://github.com/draft-js-utils/draft-js-utils/blob/master/src/Constants.js.
export const ENTITY_TYPE = {
    LINK: 'LINK',
    IMAGE: 'IMAGE',
    DOCUMENT: 'DOCUMENT',
};

// Originally from https://github.com/draft-js-utils/draft-js-utils/blob/master/src/Constants.js.
export const INLINE_STYLE = {
    BOLD: 'BOLD',
    CODE: 'CODE',
    ITALIC: 'ITALIC',
    STRIKETHROUGH: 'STRIKETHROUGH',
    UNDERLINE: 'UNDERLINE',
};

const renderMap = {};

renderMap[BLOCK_TYPE.BREAK] = { element: 'div' };

export const BlockRenderMap = Map(renderMap).merge(DefaultDraftBlockRenderMap);
