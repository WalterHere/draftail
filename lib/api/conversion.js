import { EditorState, convertFromRaw, convertToRaw } from 'draft-js';

export default {
    // Serialised RawDraftContentState => EditorState.
    createEditorState(serialisedState, decorators) {
        const rawState = JSON.parse(serialisedState);
        let editorState;

        if (rawState && Object.keys(rawState).length !== 0) {
            const contentState = convertFromRaw(rawState);
            editorState = EditorState.createWithContent(contentState, decorators);
        } else {
            editorState = EditorState.createEmpty(decorators);
        }

        return editorState;
    },

    // EditorState => Serialised RawDraftContentState.
    serialiseEditorState(editorState) {
        const contentState = editorState.getCurrentContent();
        const rawContent = convertToRaw(contentState);

        const isEmpty = rawContent.blocks.every((block) => {
            return block.text.trim().length === 0
                && block.entityRanges.length === 0
                && block.inlineStyleRanges.length === 0;
        });

        return JSON.stringify(isEmpty ? {} : rawContent);
    },
};
