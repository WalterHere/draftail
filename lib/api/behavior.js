import { Map } from 'immutable';
import {
    DefaultDraftBlockRenderMap,
    getDefaultKeyBinding,
    KeyBindingUtil,
} from 'draft-js';

import { ENTITY_TYPE, BLOCK_TYPE, INLINE_STYLE, KEY_CODES } from '../api/constants';

const { hasCommandModifier } = KeyBindingUtil;

/**
 * Behavioral methods for the editor, generated from its configuration.
 */
export default {
    /**
     * Configure block render map from block types list.
     */
    getBlockRenderMap(BLOCK_TYPES = []) {
        const renderMap = {};

        BLOCK_TYPES.filter(block => block.element)
            .forEach((block) => {
                renderMap[block.type] = { element: block.element };
            });

        return DefaultDraftBlockRenderMap.merge(Map(renderMap));
    },

    /**
     * Configure block style function from block types list.
     */
    getBlockStyleFn(BLOCK_TYPES = []) {
        const blockClassNames = {};

        BLOCK_TYPES.filter(block => block.className)
            .forEach((block) => {
                blockClassNames[block.type] = block.className;
            });

        const blockStyleFn = (block) => {
            const type = block.getType();

            return blockClassNames[type] || undefined;
        };

        return blockStyleFn;
    },

    /**
     * Configure key binding function from enabled blocks, styles, entities.
     */
    getKeyBindingFn(BLOCK_TYPES = [], INLINE_STYLES = [], ENTITY_TYPES = []) {
        const getEnabledTypes = (activeTypes, allTypes) => {
            // Go through all the possible types, and check which are enabled.
            return Object.keys(allTypes)
                .reduce((enabled, key) => {
                    // eslint-disable-next-line no-param-reassign
                    enabled[key] = activeTypes.some(item => item.type === allTypes[key]);

                    return enabled;
                }, {});
        };

        const isEnabledBlock = getEnabledTypes(BLOCK_TYPES, BLOCK_TYPE);
        const isEnabledInline = getEnabledTypes(INLINE_STYLES, INLINE_STYLE);
        const isEnabledEntity = getEnabledTypes(ENTITY_TYPES, ENTITY_TYPE);

        // Emits key commands to use in `handleKeyCommand` in `Editor`.
        const keyBindingFn = (e) => {
            // Safeguard that we only trigger shortcuts with exact matches.
            // eg. cmd + shift + b should not trigger bold.
            if (e.shiftKey) {
                // Key bindings supported by Draft.js must be explicitely discarded.
                // See https://github.com/facebook/draft-js/issues/941.
                switch (e.keyCode) {
                case KEY_CODES.B:
                    return null;
                case KEY_CODES.I:
                    return null;
                case KEY_CODES.J:
                    return null;
                case KEY_CODES.U:
                    return null;
                case KEY_CODES[5]:
                    return isEnabledInline.STRIKETHROUGH && e.shiftKey && e.altKey ? INLINE_STYLE.STRIKETHROUGH : null;
                case KEY_CODES[7]:
                    return isEnabledBlock.ORDERED_LIST_ITEM && hasCommandModifier(e) && e.shiftKey ? BLOCK_TYPE.ORDERED_LIST_ITEM : null;
                case KEY_CODES[8]:
                    return isEnabledBlock.UNORDERED_LIST_ITEM && hasCommandModifier(e) && e.shiftKey ? BLOCK_TYPE.UNORDERED_LIST_ITEM : null;
                default:
                }
            } else {
                switch (e.keyCode) {
                case KEY_CODES.K:
                    return isEnabledEntity.LINK && hasCommandModifier(e) ? ENTITY_TYPE.LINK : null;
                case KEY_CODES.B:
                    return isEnabledInline.BOLD && hasCommandModifier(e) ? INLINE_STYLE.BOLD : null;
                case KEY_CODES.I:
                    return isEnabledInline.ITALIC && hasCommandModifier(e) ? INLINE_STYLE.ITALIC : null;
                case KEY_CODES.J:
                    return isEnabledInline.CODE && hasCommandModifier(e) ? INLINE_STYLE.CODE : null;
                case KEY_CODES.U:
                    return isEnabledInline.UNDERLINE && hasCommandModifier(e) ? INLINE_STYLE.UNDERLINE : null;
                case KEY_CODES[0]:
                    // Reverting to unstyled block is always available.
                    return (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.UNSTYLED : null;
                case KEY_CODES[1]:
                    return isEnabledBlock.HEADER_ONE && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_ONE : null;
                case KEY_CODES[2]:
                    return isEnabledBlock.HEADER_TWO && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_TWO : null;
                case KEY_CODES[3]:
                    return isEnabledBlock.HEADER_THREE && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_THREE : null;
                case KEY_CODES[4]:
                    return isEnabledBlock.HEADER_FOUR && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_FOUR : null;
                case KEY_CODES[5]:
                    return isEnabledBlock.HEADER_FIVE && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_FIVE : null;
                case KEY_CODES[6]:
                    return isEnabledBlock.HEADER_SIX && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_SIX : null;
                default:
                }
            }

            return getDefaultKeyBinding(e);
        };

        return keyBindingFn;
    },

};
