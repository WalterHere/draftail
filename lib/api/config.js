import { Map } from 'immutable';
import {
    DefaultDraftBlockRenderMap,
    getDefaultKeyBinding,
    KeyBindingUtil,
} from 'draft-js';

import { ENTITY_TYPE, BLOCK_TYPE, INLINE_STYLE, KEY_CODES } from '../api/constants';

const { hasCommandModifier } = KeyBindingUtil;

// Maximum level of nesting for unordered and ordered lists.
export const MAX_LIST_NESTING = 1;
// Frequency at which the save callback is triggered.
export const STATE_SAVE_INTERVAL = 250;

/**
 * Options / configuration methods for the editor.
 */
export default {
    /**
     * Configure block render map from block types list.
     */
    getBlockRenderMap(BLOCK_TYPES = []) {
        const renderMap = {};

        BLOCK_TYPES.filter(type => type.element)
            .forEach((type) => {
                renderMap[type.style] = { element: type.element };
            });

        return DefaultDraftBlockRenderMap.merge(Map(renderMap));
    },

    /**
     * Configure block style function from block types list.
     */
    getBlockStyleFn(BLOCK_TYPES = []) {
        const blockClassNames = {};

        BLOCK_TYPES.filter(type => type.className)
            .forEach((type) => {
                blockClassNames[type.style] = type.className;
            });

        const blockStyleFn = (block) => {
            const type = block.getType();

            return blockClassNames[type] || undefined;
        };

        return blockStyleFn;
    },

    /**
     * Configure key binding function from enabled blocks, styles, entities.
     */
    getKeyBindingFn(BLOCK_TYPES = [], INLINE_STYLES = [], ENTITY_TYPES = []) {
        const isTypeEnabled = (arr, style) => !!arr.find(item => item.style === style || item.entity === style);

        const getEnabledTypes = (activeTypes, allTypes) => {
            // Go through all the possible types, and check which are enabled.
            return Object.keys(allTypes)
                .reduce((enabled, key) => {
                    // eslint-disable-next-line no-param-reassign
                    enabled[key] = isTypeEnabled(activeTypes, allTypes[key]);

                    return enabled;
                }, {});
        };

        const isEnabledBlock = getEnabledTypes(BLOCK_TYPES, BLOCK_TYPE);
        const isEnabledInline = getEnabledTypes(INLINE_STYLES, INLINE_STYLE);
        const isEnabledEntity = getEnabledTypes(ENTITY_TYPES, ENTITY_TYPE);

        // Emits key commands to use in `handleKeyCommand` in `Editor`.
        const keyBindingFn = (e) => {
            // Safeguard that we only trigger shortcuts with exact matches.
            // eg. cmd + shift + b should not trigger bold.
            if (e.shiftKey) {
                // Key bindings supported by Draft.js must be explicitely discarded.
                // See https://github.com/facebook/draft-js/issues/941.
                switch (e.keyCode) {
                case KEY_CODES.B:
                    return null;
                case KEY_CODES.I:
                    return null;
                case KEY_CODES.J:
                    return null;
                case KEY_CODES.U:
                    return null;
                case KEY_CODES[5]:
                    return isEnabledInline.STRIKETHROUGH && e.shiftKey && e.altKey ? INLINE_STYLE.STRIKETHROUGH : null;
                case KEY_CODES[7]:
                    return isEnabledBlock.ORDERED_LIST_ITEM && hasCommandModifier(e) && e.shiftKey ? BLOCK_TYPE.ORDERED_LIST_ITEM : null;
                case KEY_CODES[8]:
                    return isEnabledBlock.UNORDERED_LIST_ITEM && hasCommandModifier(e) && e.shiftKey ? BLOCK_TYPE.UNORDERED_LIST_ITEM : null;
                default:
                }
            } else {
                switch (e.keyCode) {
                case KEY_CODES.K:
                    return isEnabledEntity.LINK && hasCommandModifier(e) ? ENTITY_TYPE.LINK : null;
                case KEY_CODES.B:
                    return isEnabledInline.BOLD && hasCommandModifier(e) ? INLINE_STYLE.BOLD : null;
                case KEY_CODES.I:
                    return isEnabledInline.ITALIC && hasCommandModifier(e) ? INLINE_STYLE.ITALIC : null;
                case KEY_CODES.J:
                    return isEnabledInline.CODE && hasCommandModifier(e) ? INLINE_STYLE.CODE : null;
                case KEY_CODES.U:
                    return isEnabledInline.UNDERLINE && hasCommandModifier(e) ? INLINE_STYLE.UNDERLINE : null;
                case KEY_CODES[0]:
                    // Reverting to unstyled block is always available.
                    return (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.UNSTYLED : null;
                case KEY_CODES[1]:
                    return isEnabledBlock.HEADER_ONE && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_ONE : null;
                case KEY_CODES[2]:
                    return isEnabledBlock.HEADER_TWO && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_TWO : null;
                case KEY_CODES[3]:
                    return isEnabledBlock.HEADER_THREE && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_THREE : null;
                case KEY_CODES[4]:
                    return isEnabledBlock.HEADER_FOUR && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_FOUR : null;
                case KEY_CODES[5]:
                    return isEnabledBlock.HEADER_FIVE && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_FIVE : null;
                case KEY_CODES[6]:
                    return isEnabledBlock.HEADER_SIX && (e.ctrlKey || e.metaKey) && e.altKey ? BLOCK_TYPE.HEADER_SIX : null;
                default:
                }
            }

            // if (e.altKey === true && !e.ctrlKey) {
            //     if (e.shiftKey === true) {
            //         switch (e.which) {
            //         // Alt + Shift + A
            //         // case 65: return addNewBlock();
            //         default: return getDefaultKeyBinding(e);
            //         }
            //     }

            //     switch (e.which) {
            //     // // 1
            //     // case 49: return changeType('ordered-list-item');
            //     // // @
            //     // case 50: return showLinkInput();
            //     // // #
            //     // case 51: return changeType('header-three');
            //     // // *
            //     // case 56: return changeType('unordered-list-item');
            //     // // <
            //     // case 188: return changeType('caption');
            //     // // // -
            //     // // case 189: return 'changetype:caption';
            //     // // >
            //     // case 190: return changeType('unstyled');
            //     // // "
            //     // case 222: return changeType('blockquote');
            //     default: return getDefaultKeyBinding(e);
            //     }
            // }

            // if (e.keyCode === 46 && !e.ctrlKey) {
            //   return KEY_COMMANDS.deleteBlock();
            // }
            return getDefaultKeyBinding(e);
        };

        return keyBindingFn;
    },

};
