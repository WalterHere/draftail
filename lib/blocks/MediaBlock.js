import React, { Component } from 'react';
import { Entity } from 'draft-js';

import { ENTITY_TYPE } from '../api/constants';

import ImageBlock from '../blocks/ImageBlock';
import EmbedBlock from '../blocks/EmbedBlock';
import NullBlock from '../blocks/NullBlock';

const BLOCK_MAP = {};

BLOCK_MAP[ENTITY_TYPE.IMAGE] = ImageBlock;
BLOCK_MAP[ENTITY_TYPE.EMBED] = EmbedBlock;

/**
 * Editor block to display media and edit content.
 */
class MediaBlock extends Component {
    constructor(props) {
        super(props);

        this.state = {
            active: false,
        };

        this.toggleBlock = this.toggleBlock.bind(this);
        this.closeBlock = this.toggleBlock.bind(this, false);
        this.onClick = this.onClick.bind(this);
        this.onSave = this.onSave.bind(this);
    }

    toggleBlock(active = !this.state.active) {
        const { blockProps } = this.props;

        this.setState({
            active: active,
        });

        if (active) {
            blockProps.lockEditor();
        } else {
            blockProps.unlockEditor();
        }
    }

    onClick() {
        this.toggleBlock();
    }

    onSave(nextData) {
        const { block } = this.props;

        // This will update in place
        Entity.mergeData(block.getEntityAt(0), nextData);
        this.closeBlock();
    }

    render() {
        const { blockProps } = this.props;
        const { active } = this.state;
        const Block = BLOCK_MAP[blockProps.entity.type] || NullBlock;

        return (
            <div className={`RichEditor-media${active ? ' RichEditor-media--open' : ''}`}>
                <Block
                    entity={blockProps.entity}
                    entityConfig={blockProps.entityConfig}
                    active={active}
                    onClick={this.onClick}
                    onCancel={this.closeBlock}
                    onSave={this.onSave}
                />
            </div>
        );
    }
}

MediaBlock.propTypes = {
    block: React.PropTypes.object.isRequired,
    blockProps: React.PropTypes.object.isRequired,
};

export default MediaBlock;
