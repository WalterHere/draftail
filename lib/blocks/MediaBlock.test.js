import React from 'react';
import { shallow } from 'enzyme';
import { convertFromRaw } from 'draft-js';
import { ENTITY_TYPE } from '../api/constants';

import MediaBlock from '../blocks/MediaBlock';

const contentState = convertFromRaw({
    entityMap: {
        '0': {
            type: 'IMAGE',
            mutability: 'IMMUTABLE',
            data: {
                src:
                    'https://springload.github.io/draftail/assets/example-image.png',
            },
        },
    },
    blocks: [
        {
            key: 'b3kdk',
            text: ' ',
            type: 'atomic',
            depth: 0,
            inlineStyleRanges: [],
            entityRanges: [
                {
                    offset: 0,
                    length: 1,
                    key: 0,
                },
            ],
            data: {},
        },
    ],
});

const mockProps = {
    block: contentState.getFirstBlock(),
    blockProps: {
        entity: {
            type: 'TEST',
        },
        entityConfig: {},
    },
    contentState,
};

describe('MediaBlock', () => {
    it('basic', () => {
        expect(shallow(<MediaBlock {...mockProps} />)).toMatchSnapshot();
    });

    it('image', () => {
        expect(
            shallow(
                <MediaBlock
                    {...mockProps}
                    blockProps={{
                        entity: { type: ENTITY_TYPE.IMAGE },
                        entityConfig: {},
                    }}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('embed', () => {
        expect(
            shallow(
                <MediaBlock
                    {...mockProps}
                    blockProps={{
                        entity: { type: ENTITY_TYPE.EMBED },
                        entityConfig: {},
                    }}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('#isActive', () => {
        expect(
            shallow(<MediaBlock {...mockProps} />).setState({
                isActive: true,
            }),
        ).toMatchSnapshot();
    });

    it('#blockProps.unlockEditor', () => {
        const unlockEditor = jest.fn();
        shallow(
            <MediaBlock
                {...mockProps}
                blockProps={Object.assign(
                    { unlockEditor: unlockEditor },
                    mockProps.blockProps,
                )}
            />,
        )
            .instance()
            .onSave();
        expect(unlockEditor).toHaveBeenCalledTimes(1);
    });

    it('#blockProps.lockEditor', () => {
        const lockEditor = jest.fn();
        shallow(
            <MediaBlock
                {...mockProps}
                blockProps={Object.assign(
                    { lockEditor: lockEditor },
                    mockProps.blockProps,
                )}
            />,
        )
            .instance()
            .onClick();
        expect(lockEditor).toHaveBeenCalledTimes(1);
    });
});
