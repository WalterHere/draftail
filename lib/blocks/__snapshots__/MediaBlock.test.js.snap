exports[`MediaBlock #active 1`] = `
<div
  className="RichEditor-media RichEditor-media--open">
  <NullBlock
    active={true}
    entity={
      Object {
        "type": "TEST",
      }
    }
    entityConfig={Object {}}
    onCancel={[Function]}
    onClick={[Function]}
    onSave={[Function]} />
</div>
`;

exports[`MediaBlock basic 1`] = `
<div
  className="RichEditor-media">
  <NullBlock
    active={false}
    entity={
      Object {
        "type": "TEST",
      }
    }
    entityConfig={Object {}}
    onCancel={[Function]}
    onClick={[Function]}
    onSave={[Function]} />
</div>
`;

exports[`MediaBlock embed 1`] = `
<div
  className="RichEditor-media">
  <EmbedBlock
    active={false}
    entity={
      Object {
        "type": "EMBED",
      }
    }
    entityConfig={Object {}}
    onCancel={[Function]}
    onClick={[Function]}
    onSave={[Function]} />
</div>
`;

exports[`MediaBlock image 1`] = `
<div
  className="RichEditor-media">
  <ImageBlock
    active={false}
    entity={
      Object {
        "type": "IMAGE",
      }
    }
    entityConfig={Object {}}
    onCancel={[Function]}
    onClick={[Function]}
    onSave={[Function]} />
</div>
`;
